Multi Objective CNN benchmark - [WIP]

ayushi-3536 opened the pull request:
- Added benchmark outline for CNN benchmarks from [paper](https://arxiv.org/abs/2105.01015)
- Pull from MO interface by Phillip is pending

ayushi-3536 committed:
-added mo cnn benchmarks from bag of baseline paper
- datamanager in progress

ayushi-3536 committed:
-removing log from categorical cs

ayushi-3536 committed:
- integrating data loaded output
- preprocessing of data

Commit:
integration of data manager with bench;changes in cs paramname

ayushi-3536 committed:
- removing unwanted logs

PhMueller commented on hpobench/benchmarks/mo/cnn_benchmark.py:
```diff
@@ -0,0 +1,535 @@
+        if isinstance(config, CS.Configuration):
+            config = config.get_dictionary()
+        if isinstance(fidelity, CS.Configuration):
+            fidelity = config.get_dictionary()
```
Passing the fidelity as argument seems have no effect. (not returned afterwards)

PhMueller commented on hpobench/benchmarks/mo/cnn_benchmark.py:
```diff
@@ -0,0 +1,535 @@
+            fidelity = config.get_dictionary()
+        return Net(config, (3, 16, 16), 17)
+
+    @AbstractBenchmark.check_parameters
```
Please use 

```suggestion
    @AbstractMultiObjectiveBenchmark.check_parameters
```
here. Also, please for the objective function test.

PhMueller commented on hpobench/benchmarks/mo/cnn_benchmark.py:
```diff
@@ -0,0 +1,535 @@
+        return acc
+
+
+class CNNBenchmark(AbstractBenchmark):
```
After rebasing to the new state, we can use the MO interface

```suggestion
class CNNBenchmark(AbstractMultiObjectiveBenchmark):
```

PhMueller commented on hpobench/benchmarks/mo/cnn_benchmark.py:
```diff
@@ -0,0 +1,535 @@
+        return cs
+
+    @staticmethod
+    def get_objectives():
```
```suggestion
    def get_objective_names(self) -> List[str]:
```

PhMueller commented on hpobench/benchmarks/mo/cnn_benchmark.py:
```diff
@@ -0,0 +1,535 @@
+        )
+        t.close()
+
+        return {'function_value': {'accuracy': val_accuracy,
```
Can you please make sure that the returned vales are python objects. Non-python objects might cause some problems when using with the container. 

Example: `val_accuracy.item()`

PhMueller commented on hpobench/benchmarks/mo/cnn_benchmark.py:
```diff
@@ -0,0 +1,535 @@
+                         }
+                }
+
+    @AbstractBenchmark.check_parameters
```
```suggestion
    @AbstractMultiObjectiveBenchmark.check_parameters
```

PhMueller commented on hpobench/util/data_manager.py:
```diff
@@ -845,6 +844,89 @@ def _load(self) -> Tuple[np.ndarray, np.ndarray, np.ndarray, np.ndarray, np.ndar
+        return X_trn, y_trn, X_val, y_val, X_tst, y_tst
+
+    @lockutils.synchronized('not_thread_process_safe', external=True,
+                            lock_path=f'{hpobench.config_file.cache_dir}/lock_protein_structure_data', delay=0.5)
```
can you give the lock please another name?

ayushi-3536 committed:
Merge branch 'development' of github.com:automl/HPOBench into mo_cnn

ayushi-3536 committed:
- added singularity config for mo cnn benches

ayushi-3536 committed:
add \ to the end of command

ayushi-3536 committed:
remove redundant command

ayushi-3536 committed:
cleanup

ayushi-3536 committed:
- rebase mo interface
- changed lock name for mo_cnn bench

ayushi-3536 committed:
- removing sample size fidelity
- removed hard coded model input to support multiple datasets

ayushi-3536 committed:
Revert "- removing sample size fidelity"

This reverts commit f76ea99d5de17a6bddb51328557f3764e3dab466.

ayushi-3536 committed:
- removed hard coded input size and output class to facilitate integration of various datasets
- changed epoch from 50 to 25 (from literature)
- corrected epoch training(0 indexed)
- removed subsample from fidelity(not done in literature, can discuss to add it if we want to perform experiments for this)
- returning python object

PhMueller commented on hpobench/benchmarks/mo/cnn_benchmark.py:
```diff
@@ -0,0 +1,520 @@
+        layers.append(nn.Linear(inp_n, num_classes))
+        self.fc_layers = nn.Sequential(*layers)
+
+        self.time_train = 0
```
This attribute seems to be unused

PhMueller commented on hpobench/benchmarks/mo/cnn_benchmark.py:
```diff
@@ -0,0 +1,520 @@
+        return cs
+
+    @staticmethod
+    def get_objective_names(self) -> List[str]:
```
```suggestion
    def get_objective_names() -> List[str]:
```

PhMueller commented on hpobench/benchmarks/mo/cnn_benchmark.py:
```diff
@@ -0,0 +1,520 @@
+        return acc
+
+    def eval_fn(self, loader, device):
+        """
```
Could you please use the google style for docstrings.

PhMueller commented on hpobench/benchmarks/mo/cnn_benchmark.py:
```diff
@@ -0,0 +1,520 @@
+
+        budget = fidelity1[iter_choice]
+        return budget
+
```
We could merge this function with the `get_fidelity_space` function to increase readability.

PhMueller commented on hpobench/benchmarks/mo/cnn_benchmark.py:
```diff
@@ -0,0 +1,520 @@
+                           'author  = {Guerrero-Viu, Julia and Hauns, Sven and Izquierdo, Sergio and Miotto, Guilherme and Schrodi, Simon and Biedenkapp, Andre and Elsken, Thomas and Deng, Difan and Lindauer, Marius and Hutter, Frank},},'
+                           'journal = {arXiv preprint arXiv:2105.01015},'
+                           'year    = {2021}}',
+                           ],
```
Could you please add line breaks, so that the 120 chars per line limit is fulfilled.

PhMueller replied:
in general, also please check the code with;

``` bash
flake8 --max-line-length=120 ./hpobench
pycodestyle --max-line-length=120 ./hpobench
pylint --disable=all --enable=W0221 ./hpobench
```

PhMueller commented on hpobench/benchmarks/mo/cnn_benchmark.py:
```diff
@@ -0,0 +1,520 @@
+                   rng: Union[int, np.random.RandomState, None] = None):
+        """ Function that returns the model initialized based on the configuration and fidelity
+        """
+        rng = self.rng if rng is None else rng
```
`rng` is not used

PhMueller commented on hpobench/util/data_manager.py:
```diff
@@ -845,6 +844,89 @@ def _load(self) -> Tuple[np.ndarray, np.ndarray, np.ndarray, np.ndarray, np.ndar
+        return X_trn, y_trn, X_val, y_val, X_tst, y_tst
+
+    @lockutils.synchronized('not_thread_process_safe', external=True,
+                            lock_path=f'{hpobench.config_file.cache_dir}/lock_mo_cnn_data', delay=0.5)
```
Can you please remove this for now. 

We have already some generic functions for the dataloading. But since they create their own lock, we dont need that here. 

(I'll make a rework of the data manager soon. )

PhMueller replied:
```suggestion
```

PhMueller commented on hpobench/util/data_manager.py:
```diff
@@ -845,6 +844,89 @@ def _load(self) -> Tuple[np.ndarray, np.ndarray, np.ndarray, np.ndarray, np.ndar
+            self.compressed_data.parent.mkdir(parents=True, exist_ok=True)
+            urlretrieve(self.url_source, self.compressed_data)
+            tar = tarfile.open(self.compressed_data)
+            tar.extractall(self.save_dir)
```
See comment above. 
Also just as a tip, it is always good to use `file.open(..)` inside a context manager (`with tarfile.open(..)`) otherwise you have to close it afterwards manually (`tarfile.close()`)

```suggestion
            self._download_file_with_progressbar(data_url=self.url_source, data_file=self.compressed_data)
            self._untar_data(compressed_file=self.compressed_data, save_dir=self.save_dir)
```

PhMueller commented on hpobench/benchmarks/mo/cnn_benchmark.py:
```diff
@@ -0,0 +1,520 @@
+            t.update()
+        training_runtime = time.time() - start
+
+        num_params = np.sum(p.numel() for p in model.parameters())
```
```suggestion
        num_params = np.sum([p.numel() for p in model.parameters()])
```

PhMueller commented on hpobench/benchmarks/mo/cnn_benchmark.py:
```diff
@@ -0,0 +1,520 @@
+            To prevent overfitting on a single seed, it is possible to pass a
+            parameter ``rng`` as 'int' or 'np.random.RandomState' to this function.
+            If this parameter is not given, the default random state is used.
+
```
Add docstring for `shuffle` please

PhMueller commented on hpobench/benchmarks/mo/cnn_benchmark.py:
```diff
@@ -0,0 +1,520 @@
+            Fidelity parameters, check get_fidelity_space(). Uses default (max) value if None.
+        rng : np.random.RandomState, int, None
+            Random seed to use in the benchmark.
+
```
see above

PhMueller commented on hpobench/benchmarks/mo/cnn_benchmark.py:
```diff
@@ -0,0 +1,520 @@
+            t.update()
+        training_runtime = time.time() - start
+
+        num_params = np.sum(p.numel() for p in model.parameters())
```
```suggestion
        num_params = np.sum([p.numel() for p in model.parameters()])
```

PhMueller commented on hpobench/container/benchmarks/mo/cnn_benchmark.py:
```diff
@@ -0,0 +1,20 @@
+    def __init__(self, **kwargs):
+        kwargs['benchmark_name'] = kwargs.get('benchmark_name', 'FlowerCNNBenchmark')
+        kwargs['container_name'] = kwargs.get('container_name', 'mo_cnn')
+        kwargs['latest'] = kwargs.get('container_tag', '0.0.4')
```
I think it should be 0.0.1, right?

PhMueller commented on hpobench/container/benchmarks/mo/cnn_benchmark.py:
```diff
@@ -0,0 +1,20 @@
+    def __init__(self, **kwargs):
+        kwargs['benchmark_name'] = kwargs.get('benchmark_name', 'FashionCNNBenchmark')
+        kwargs['container_name'] = kwargs.get('container_name', 'mo_cnn')
+        kwargs['latest'] = kwargs.get('container_tag', '0.0.4')
```
same here

PhMueller committed:
Incorporate Requests

PhMueller committed:
Make deterministic

PhMueller committed:
Add test file

PhMueller committed:
Fix Test Config

ayushi-3536 committed:
- report total elapsed time in obj func as cost(as in original bench)
- merged fidelity space and choice method

PhMueller committed:
Update Recipe

PhMueller committed:
Add init and log statement

PhMueller committed:
Add __init__ files

PhMueller committed:
Clean up

PhMueller committed:
Allow to evaluate the test performance on different budgets

PhMueller committed:
Enable the gpu automatically for the container.

PhMueller committed:
Adapt Test

KEggensperger commented on extra_requirements/mo_cnn.json:
```diff
@@ -0,0 +1,7 @@
+{
+  "mo_cnn": [
+    "tqdm",
```
version?

KEggensperger commented on hpobench/benchmarks/mo/cnn_benchmark.py:
```diff
@@ -0,0 +1,591 @@
+
+        return acc
+
+    def eval_fn(self, loader, device):
```
could you please add types for all signatures?

KEggensperger commented on hpobench/benchmarks/mo/cnn_benchmark.py:
```diff
@@ -0,0 +1,591 @@
+        rng = self.rng if rng is None else rng
+        if isinstance(config, CS.Configuration):
+            config = config.get_dictionary()
+        return Net(config, self.input_shape, self.output_classes)
```
Shouldn't this take the rng?

KEggensperger commented on hpobench/benchmarks/mo/cnn_benchmark.py:
```diff
@@ -0,0 +1,591 @@
+        val_accuracy = model.eval_fn(ds_val, device).item()
+        eval_valid_runtime = time.time() - start
+        start = time.time()
+        test_accuracy = model.eval_fn(ds_test, device).item()
```
Same questions as for the other benchmark: Why spending time on computing test metrics?

PhMueller replied:
Good question. Changed it to "training time". 

The eval time should be almost equal for every run, so i think it is more important to report the "training time" instead of the "total time per configuration". 

Thanks for the feedback!

PhMueller committed:
Add missing signatures

PhMueller committed:
Return training time instead of evaluation time.

PhMueller committed:
Fix dependency version

PhMueller committed:
Update CNN Mo Tests

ayushi-3536 committed:
-add conditional dependencies in search space
-update test case

PhMueller committed:
Update tests

PhMueller committed:
Merge remote-tracking branch 'fork_ayushi/mo_cnn' into mo_cnn

PhMueller committed:
Update tests

PhMueller committed:
Simlify ConfigSpace. + Update Tests.

PhMueller committed:
Simlify ConfigSpace. + Update Tests.

PhMueller committed:
Simlify ConfigSpace. + Update Tests.

PhMueller committed:
Change returned Objectives

PhMueller committed:
Change returned Objectives.. again

PhMueller committed:
Change inheritance in container benchmark
